### PR TITLE
Improve overflow detection

### PR DESCRIPTION
**Pull request by Flamefire**
This stores an ever-increasing number maxNumProc in addition to the current next id. This is the maximum number of processes ever used to generate ids. With this, an overflow can be detected on all ranks in cases where it was not possible before.

Due to this change the get/setNextId was changed to get/setState which is more appropriate.

ccing @psychocoderHPC

### CODE REVIEW

**Comment by Flamefire** [comment]
@psychocoderHPC This now works as it should and can be merged

File: src/libPMacc/include/particles/IdProvider.hpp [hunk 1]
```diff
@@ -55,29 +56,40 @@ namespace PMacc {
...
         const uint64_t globalUniqueStartId = getStartId();
-        setNextId(globalUniqueStartId);
+        maxNumProc_ = Environment<T_dim>::get().GridController().getGpuNodes().productOfComponents();
+        // Init value to avoid uninitialized read warnings
+        setState(globalUniqueStartId, 0);
         // Instantiate kernel
         getNewIdHost();
         // Reset to start value
-        setNextId(globalUniqueStartId);
+        setState(globalUniqueStartId, maxNumProc_);
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/IdProvider.hpp]: please use a name without leading `_`

File: src/libPMacc/include/particles/IdProvider.hpp [hunk 2]
```diff
@@ -94,26 +106,33 @@ namespace PMacc {
...
+        // Number of bits in the ids
+        BOOST_STATIC_CONSTEXPR int32_t numBitsOfType = sizeof(maxNumProc_) * CHAR_BIT;
+
+        // Get current value
+        uint64_t nextId = getState().get<0>();
+        /* Example: maxNumProc_ has 3 set bits (<8 ranks), 64bit value used
+         * --> Shift by 61 bits
+         * => 3 upper bits are left untouched (besides moving), rest is zero
+         */
+        nextId >>= numBitsOfType - bitsToCheck;
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/IdProvider.hpp]: offline discussed: 
- the current implementation is wrong
- we need the start id

**Comment by ax3l** [comment]
@psychocoderHPC updated :)

**Comment by psychocoderHPC** [comment]
Thanks for this great pull request.

**Pull request merged**